### PR TITLE
Xcode 8 & Swift 3 migration

### LINKED ISSUE #49: Migrate to Swift 3.0

### PR DESCRIPTION
**Pull request by diogoguimaraes**
Closes #49.
Also the project's structure has been updated to support SPM in the future.

### CODE REVIEW

**Comment by SandroMachado** [comment]
- [ ] Please make sure that all the models properties are `private` and not `fileprivate` to allow use from the enclosing declaration.

**Review by SandroMachado — DISMISSED** [review]
- [ ] Please make sure that all the models properties are `private` and not `fileprivate` to allow use from the enclosing declaration.

File: Source/Model/Transaction.swift [hunk 1]
```diff
@@ -182,7 +182,7 @@ public class Transaction: BaseModel, Mappable {
 
       - returns: A promise with the transaction.
     */
-    public func commit(otp: String) -> Promise<Transaction> {
+    open func commit(_ otp: String) -> Promise<Transaction> {
```
Comment (SandroMachado) [inline Source/Model/Transaction.swift]: Remove the `_` and add the label to the `commit`. Do the same for the other methods.

File: Source/Model/User.swift [hunk 2]
```diff
@@ -98,7 +98,7 @@ public class User: BaseModel, Mappable {
 
       - returns: A promise with the created card.
     */
-    public func createCard(cardRequest: CardRequest) -> Promise<Card> {
+    open func createCard(_ cardRequest: CardRequest) -> Promise<Card> {
```
Comment (SandroMachado) [inline Source/Model/User.swift]: Please see my comment above.

File: Source/Model/User.swift [hunk 3]
```diff
@@ -111,7 +111,7 @@ public class User: BaseModel, Mappable {
 
       - returns: A promise with the created contact.
      */
-    public func createContact(contactRequest: ContactRequest) -> Promise<Contact> {
+    open func createContact(_ contactRequest: ContactRequest) -> Promise<Contact> {
```
Comment (SandroMachado) [inline Source/Model/User.swift]: Please see my comment above.

File: Source/Model/User.swift [hunk 4]
```diff
@@ -135,7 +135,7 @@ public class User: BaseModel, Mappable {
 
       - returns: A promise with the user account.
     */
-    public func getAccountById(accountId: String) -> Promise<Account> {
+    open func getAccountById(_ accountId: String) -> Promise<Account> {
```
Comment (SandroMachado) [inline Source/Model/User.swift]: Please see my comment above.

File: Source/Model/User.swift [hunk 5]
```diff
@@ -182,7 +182,7 @@ public class User: BaseModel, Mappable {
 
       - returns: A promise with the user balance for the currency.
      */
-    public func getBalanceByCurrency(currency: String) -> Promise<Currency> {
+    open func getBalanceByCurrency(_ currency: String) -> Promise<Currency> {
```
Comment (SandroMachado) [inline Source/Model/User.swift]: Please see my comment above.

File: Source/Model/User.swift [hunk 6]
```diff
@@ -229,7 +229,7 @@ public class User: BaseModel, Mappable {
 
       - returns: A promise with the card.
      */
-    public func getCardById(cardId: String) -> Promise<Card> {
+    open func getCardById(_ cardId: String) -> Promise<Card> {
```
Comment (SandroMachado) [inline Source/Model/User.swift]: Please see my comment above.

File: Source/Model/User.swift [hunk 7]
```diff
@@ -242,7 +242,7 @@ public class User: BaseModel, Mappable {
 
       - returns: A promise with the list of cards.
      */
-    public func getCardsByCurrency(currency: String) -> Promise<[Card]> {
+    open func getCardsByCurrency(_ currency: String) -> Promise<[Card]> {
```
Comment (SandroMachado) [inline Source/Model/User.swift]: Please see my comment above.

File: Source/Service/ReserveService.swift [hunk 8]
```diff
@@ -33,7 +33,7 @@ public class ReserveService {
 
       - returns: A request to get the reserve transactions.
     */
-    static func getReserveTransactions(range: String) -> Request {
+    static func getReserveTransactions(_ range: String) -> Request {
```
Comment (SandroMachado) [inline Source/Service/ReserveService.swift]: Please see my comment above.

File: Source/Service/ReserveService.swift [hunk 9]
```diff
@@ -22,7 +22,7 @@ public class ReserveService {
 
       - returns: A request to get a reserve transaction.
     */
-    static func getReserveTransactionById(transactionId: String) -> Request {
+    static func getReserveTransactionById(_ transactionId: String) -> Request {
```
Comment (SandroMachado) [inline Source/Service/ReserveService.swift]: Please see my comment above.

File: Source/Service/TickerService.swift [hunk 10]
```diff
@@ -20,7 +20,7 @@ public class TickerService {
 
       - returns: A request to get the tickers on the system for the selected currency.
     */
-    static func getAllTickersByCurrency(currency: String) -> Request {
+    static func getAllTickersByCurrency(_ currency: String) -> Request {
```
Comment (SandroMachado) [inline Source/Service/TickerService.swift]: Please see my comment above.

File: Source/Service/UserCardService.swift [hunk 11]
```diff
@@ -12,7 +12,7 @@ public class UserCardService {
 
       - returns: A request to cancel the transaction.
     */
-    static func cancelTransaction(cardId: String, transactionId: String) -> Request {
+    static func cancelTransaction(_ cardId: String, transactionId: String) -> Request {
```
Comment (SandroMachado) [inline Source/Service/UserCardService.swift]: Please see my comment above.

File: Source/Service/UserCardService.swift [hunk 12]
```diff
@@ -24,7 +24,7 @@ public class UserCardService {
 
       - returns: A request to create a card.
     */
-    static func createCardAddress(cardId: String, addressRequest: AnyObject) -> Request {
+    static func createCardAddress(_ cardId: String, addressRequest: Any) -> Request {
```
Comment (SandroMachado) [inline Source/Service/UserCardService.swift]: Please see my comment above.

File: Source/Service/UserCardService.swift [hunk 13]
```diff
@@ -38,10 +38,10 @@ public class UserCardService {
 
       - returns: A request to confirm the transaction.
     */
-    static func confirmTransaction(cardId: String, otp: String?, transactionId: String, transactionCommitRequest: AnyObject?) -> Request {
+    static func confirmTransaction(_ cardId: String, otp: String?, transactionId: String, transactionCommitRequest: Any?) -> Request {
```
Comment (SandroMachado) [inline Source/Service/UserCardService.swift]: Please see my comment above.

File: Source/Service/UserCardService.swift [hunk 14]
```diff
@@ -63,7 +63,7 @@ public class UserCardService {
 
       - returns: A request to create a transaction.
     */
-    static func createTransaction(cardId: String, commit: Bool, transactionRequest: AnyObject) -> Request {
+    static func createTransaction(_ cardId: String, commit: Bool, transactionRequest: Any) -> Request {
```
Comment (SandroMachado) [inline Source/Service/UserCardService.swift]: Please see my comment above.

File: Source/Util/Header.swift [hunk 15]
```diff
@@ -75,12 +75,12 @@ public class Header {
...
       - returns: The total number of elements.
     */
-    public static func getTotalNumberOfResults(headers: [String: String]) -> Int? {
+    open static func getTotalNumberOfResults(_ headers: [String: String]) -> Int? {
         guard let contentRange = headers["content-range"] else {
             return nil
         }
 
-        return NSString(string: contentRange.substringFromIndex(contentRange.rangeOfString("/")!.endIndex)).integerValue
+        return NSString(string: contentRange.substring(from: contentRange.range(of: "/")!.upperBound)).integerValue
```
Comment (SandroMachado) [inline Source/Util/Header.swift]: Remove the force unwrap.

File: Tests/Integration/Client/UpholdClientTest.swift [hunk 16]
```diff
@@ -204,10 +210,10 @@ class UpholdClientTest: UpholdTestCase {
                 "}" +
             "}" +
         "}"
-        let client = UpholdClient(bearerToken: "foobar")
-        client.token.adapter = MockRestAdapter(body: json)
+        let client = MockUpholdClient(bearerToken: "foobar")
+        client.mockToken.mockAdapter = MockRestAdapter(body: json)
 
-        client.getUser().then { (user: User) -> () in
+        _ = client.getUser().then { (user: User) -> () in
```
Comment (SandroMachado) [inline Tests/Integration/Client/UpholdClientTest.swift]: Improve this declaration:
```
let user = client.getUser();
user.then { (user: User) -> () in...
```

**Review by d-moreira — APPROVED** [review]

**Review by ruipenso — APPROVED** [review]

**Pull request merged**